builtin/credential/approle/path_tidy_user_id.go

```diff
@@ -165,21 +165,27 @@ func (b *backend) tidySecretID(ctx context.Context, req *logical.Request) (*logi
 			// Accessor indexes were not getting cleaned up until 0.9.3. This is a fix
 			// to clean up the dangling accessor entries.
 			if len(accessorMap) > 0 {
+				logger.Trace(fmt.Sprintf("dangling accessorMap length: %d", len(accessorMap)))
+				// Wait to get a lock for each of the secretIDs to make sure we avoid a race conditon
+				// where the accessor has been written but not yet the secret
 				for _, lock := range b.secretIDLocks {
 					lock.Lock()
-					defer lock.Unlock()
+					lock.Unlock()
```

Comment:
Is this in the right place? It seems strange to me that we would grab a lock and then unlock it, plus below there's a comment that we're doing stuff now that we waited for the lock. Usually locks are held for the duration of time when we're doing work that might otherwise lead to a race condition.

Reply:
I can't say for certain if this is the right place. But for the current implementation I believe this does achieve the same goal with way less downtime. 

> so if we have the lock we know we're not in a wrote-accessor-but-not-yet-secret case, which can be racy.

So this means we want to make sure that nothing is currently still writing the secret-id between these two calls.

First a list of accessors is created. It is possible that an accessor exists but the secretID has not been written yet. 

https://github.com/hashicorp/vault/blob/6e96d87d5e9e57821dceb4ca8205d1965125fa94/builtin/credential/approle/path_tidy_user_id.go#L71

Then a list of secretIDs is made:

https://github.com/hashicorp/vault/blob/6e96d87d5e9e57821dceb4ca8205d1965125fa94/builtin/credential/approle/path_tidy_user_id.go#L153

In the scenario where the secretID does not exist yet this accessor will not be deleted and will be considered dangling. 

By waiting for a lock here:

https://github.com/hashicorp/vault/blob/6e96d87d5e9e57821dceb4ca8205d1965125fa94/builtin/credential/approle/path_tidy_user_id.go#L171-L186

We can be sure that all secretIDs have been written since the list of accessors was created. Which prevents the race condition from occurring. 

After stepping through this I think I understand your comment better and realise that all of this locking could occur earlier without any need to do later "cleanups". 

Waiting for the lock (and immediately releasing it) here:

https://github.com/hashicorp/vault/blob/6e96d87d5e9e57821dceb4ca8205d1965125fa94/builtin/credential/approle/path_tidy_user_id.go#L151-L163

Would achieve the same thing without needing to do the cleanup for dangling accessors. By doing it here I believe that the accessor could be immediately removed once being detected as being marked as dangling and that https://github.com/hashicorp/vault/blob/6e96d87d5e9e57821dceb4ca8205d1965125fa94/builtin/credential/approle/path_tidy_user_id.go#L167 could be completely removed altogether as the situation would no longer be possible. 

I'll test this locally first and see if it still prevents the race condition.

Reply:
Yup, looks like it works just fine! I just pushed a new commit that waits for the lock before listing the secretIDHMACs. When commenting out the waiting for the lock I can get a race condition within around 30 seconds. With the current code in place I have run it for 40 minutes (4 in parallel for 10 minutes) without hitting a single race condition.